src/twisted/application/runner/_pidfile.py

```diff
@@ -0,0 +1,270 @@
+# -*- test-case-name: twisted.application.runner.test.test_pidfile -*-
+# Copyright (c) Twisted Matrix Laboratories.
+# See LICENSE for details.
+
+"""
+PID file.
+"""
+
+import errno
+from os import getpid, kill
+
+from zope.interface import Interface, implementer
+
+from twisted.logger import Logger
+
+
+
+class IPIDFile(Interface):
+    """
+    PID file.
+
+    Manages a file that remembers a process ID.
+    """
+
+    def read():
+        """
+        Read the process ID stored in this PID file.
+
+        @return: The contained process ID.
+        @rtype: L{int}
+
+        @raise NoPIDFound: If this PID file does not exist.
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise ValueError: If this PID file's content is invalid.
+        """
+
+
+    def write(pid):
+        """
+        Store a PID in this PID file.
+
+        @param pid: A PID to store.
+        @type pid: L{int}
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+
+
+    def writeRunningPID():
+        """
+        Store the PID of the current process in this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be written.
+        """
+
+
+    def remove():
+        """
+        Remove this PID file.
+
+        @raise EnvironmentError: If this PID file cannot be removed.
+        """
+
+
+    def isRunning():
+        """
+        Determine whether there is a running process corresponding to the PID
+        in this PID file.
+
+        @return: True if this PID file contains a PID and a process with that
+        PID is currently running; false otherwise.
+        @rtype: L{bool}
+
+        @raise EnvironmentError: If this PID file cannot be read.
+        @raise InvalidPIDFileError: If this PID file's content is invalid.
+        @raise StalePIDFileError: If this PID file's content refers to a PID
+            for which there is no corresponding running process.
+        """
+
+
+    def __enter__():
+        """
+        Enter a context using this PIDFile.
+
+        Writes the PID file with the PID of the running process.
+
+        @raise AlreadyRunningError: A process corresponding to the PID in this
+            PID file is already running.
+        """
+
+
+    def __exit__(exc_type, exc_value, traceback):
+        """
+        Exit a context using this PIDFile.
+
+        Removes the PID file.
+        """
+
+
+
+@implementer(IPIDFile)
+class PIDFile(object):
+    """
+    Concrete implementation of L{IPIDFile} based on C{IFilePath}.
+    """
+
+    _log = Logger()
+
+
+    @staticmethod
+    def format(pid):
```

Comment:
We should avoid exporting any public API surface that we don't need; in this case, `PIDFile`'s job should just concretely implementing `IPIDFile` and should avoid exporting any other public methods.  This `format` implementation, for example, is tightly coupled to the `read` implementation, and substituting it, invoking it, or overriding it would not actually work without dealing with that dependency.